syntax = "proto3";

package eventstore;

service EventStore {
  rpc publish (PublishRequest) returns (PublishResponse) {}
  rpc subscribe (SubscribeRequest) returns (stream Notification) {}
  rpc unsubscribe (UnsubscribeRequest) returns (UnsubscribeResponse) {}
  rpc get_all (GetAllRequest) returns (GetAllResponse) {}
}

message PublishRequest {
  string event_topic = 1;
  string event_data = 2;
}

message PublishResponse {
  string entry_id = 1;
}

message SubscribeRequest {
  string event_topic = 1;
}

message Notification {
  string event_id = 1;
  double event_ts = 2;
  string event_action = 3;
  string event_data = 4;
}

message UnsubscribeRequest {
  string event_topic = 1;
  string event_action = 2;
}

message UnsubscribeResponse {
  bool success = 1;
}

message GetAllRequest {
  string event_topic = 1;
}

message GetAllResponse {
  string events = 1;
}
